src/persist-client/src/cli/admin.rs

```diff
@@ -273,3 +305,158 @@ pub async fn force_compaction(
+        Ok(None)
+    }
+}
+
+#[async_trait]
+impl Consensus for ReadOnly<Arc<dyn Consensus + Sync + Send>> {
+    async fn head(&self, key: &str) -> Result<Option<VersionedData>, ExternalError> {
+        self.0.head(key).await
+    }
+
+    async fn compare_and_set(
+        &self,
+        _key: &str,
+        _expected: Option<SeqNo>,
+        _new: VersionedData,
+    ) -> Result<Result<(), Vec<VersionedData>>, ExternalError> {
+        Ok(Ok(()))
+    }
+
+    async fn scan(
+        &self,
+        key: &str,
+        from: SeqNo,
+        limit: usize,
+    ) -> Result<Vec<VersionedData>, ExternalError> {
+        self.0.scan(key, from, limit).await
+    }
+
+    async fn truncate(&self, _key: &str, _seqno: SeqNo) -> Result<usize, ExternalError> {
+        Ok(0)
+    }
+}
+
+async fn force_gc(
+    cfg: PersistConfig,
+    metrics_registry: &MetricsRegistry,
+    shard_id: ShardId,
+    consensus_uri: &str,
+    blob_uri: &str,
+    commit: bool,
+) -> anyhow::Result<Box<dyn Any>> {
+    let metrics = Arc::new(Metrics::new(&cfg, metrics_registry));
+    let consensus = ConsensusConfig::try_from(
+        consensus_uri,
+        Box::new(cfg.clone()),
+        metrics.postgres_consensus.clone(),
+    )?;
+    let consensus = consensus.clone().open().await?;
+    let consensus = if commit {
+        consensus
+    } else {
+        Arc::new(ReadOnly(consensus))
+    };
+    let consensus: Arc<dyn Consensus + Send + Sync> =
+        Arc::new(MetricsConsensus::new(consensus, Arc::clone(&metrics)));
+    let blob = BlobConfig::try_from(blob_uri).await?;
+    let blob = blob.clone().open().await?;
+    let blob = if commit {
+        blob
+    } else {
+        Arc::new(ReadOnly(blob))
+    };
+    let blob: Arc<dyn Blob + Send + Sync> = Arc::new(MetricsBlob::new(blob, Arc::clone(&metrics)));
+
+    let state_versions = Arc::new(StateVersions::new(
+        cfg.clone(),
+        consensus,
+        Arc::clone(&blob),
+        Arc::clone(&metrics),
+    ));
+
+    // Prime the K V codec magic
+    let versions = state_versions
+        .fetch_recent_live_diffs::<u64>(&shard_id)
+        .await;
+
+    loop {
+        let state_res = state_versions
+            .fetch_current_state::<crate::cli::inspect::K, crate::cli::inspect::V, u64, i64>(
+                &shard_id,
+                versions.0.clone(),
+            )
+            .await;
+        let state = match state_res {
+            Ok(state) => state,
+            Err(codec) => {
+                let mut kvtd = crate::cli::inspect::KVTD_CODECS.lock().expect("lockable");
+                *kvtd = codec.actual;
+                continue;
+            }
+        };
+        // This isn't the perfect place to put this check, the ideal would be in
+        // the apply_unbatched_cmd loop, but I don't want to pollute the prod
+        // code with this logic.
+        let safe_version_change = if commit {
+            cfg.build_version == state.applier_version
```

Comment:
I'm not familiar with whether cherry-picking this onto a release branch gets exactly the release version, though this check should prevent us from messing things up too badly if we get it wrong.

Reply:
it's easy to check. this version is just whatever is in src/persist-client/Cargo.toml